js/pro/poloniex.js

```diff
@@ -0,0 +1,886 @@
+                'watchOHLCV': true,
+                'watchOrderBook': true,
+                'watchTicker': true,
+                'watchTickers': true,
+                'watchTrades': true,
+                'watchBalance': true,
+                'watchStatus': false,
+                'watchOrders': true,
+                'watchMyTrades': false,
+            },
+            'urls': {
+                'api': {
+                    'ws': {
+                        'public': 'wss://ws.poloniex.com/ws/public',
+                        'private': 'wss://ws.poloniex.com/ws/private',
+                    },
+                },
+            },
+            'options': {
+                'tradesLimit': 1000,
+                'ordersLimit': 1000,
+                'OHLCVLimit': 1000,
+                'connectionsLimit': 2000, // 2000 public, 2000 private, 4000 total, only for subscribe events, unsubscribe not restricted
+                'requestsLimit': 500, // per second, only for subscribe events, unsubscribe not restricted
+                'channelToTimeframe': {
+                    'candles_minute_1': '1m',
+                    'candles_minute_5': '5m',
+                    'candles_minute_10': '10m',
+                    'candles_minute_15': '15m',
+                    'candles_minute_30': '30m',
+                    'candles_hour_1': '1h',
+                    'candles_hour_2': '2h',
+                    'candles_hour_4': '4h',
+                    'candles_hour_6': '6h',
+                    'candles_hour_12': '12h',
+                    'candles_day_1': '1d',
+                    'candles_day_3': '3d',
+                    'candles_week_1': '1w',
+                    'candles_month_1': '1m',
+                },
+            },
+        });
+    }
+
+    async authenticate () {
+        /**
+         * @ignore
+         * @method
+         * @description authenticates the user to access private web socket channels
+         * @see https://docs.poloniex.com/#authenticated-channels-market-data-authentication
+         * @returns {object} response from exchange
+         */
+        this.checkRequiredCredentials ();
+        const timestamp = this.milliseconds ();
+        const accessPath = '/ws/private';
+        const requestString = 'GET\n' + accessPath + '\nsignTimestamp=' + timestamp;
+        const url = this.urls['api']['ws']['private'];
+        const signature = this.hmac (this.encode (requestString), this.base64ToBinary (this.secret), 'sha256', 'base64');
+        const request = {
+            'event': 'subscribe',
+            'channel': [ 'auth' ],
+            'params': {
+                'key': this.apiKey,
+                'signTimestamp': timestamp,
+                'signature': signature,
+                'signatureMethod': 'HmacSHA256',  // optional
+                'signatureVersion': '2',          // optional
+            },
+        };
+        return await this.watch (url, accessPath, request, accessPath);
```

Comment:
@samgermain this does not seem right, we should only authenticate once, and after that we should just return the resolved futures, so if we do this:

```
await this.authenticate() // authenticates
await this.authenticate() // just returns the future
await this.authenticate() // just returns the future
await this.authenticate() // just returns the future
```
Please check how others exchanges do this.

Reply:
> @samgermain this does not seem right, we should only authenticate once, and after that we should just return the resolved futures, so if we do this:
> 
> ```
> await this.authenticate() // authenticates
> await this.authenticate() // just returns the future
> await this.authenticate() // just returns the future
> await this.authenticate() // just returns the future
> ```
> 
> Please check how others exchanges do this.

Other exchanges look like they implement it the same way, in that authenticate is called every time the watch method is called.

https://github.com/samgermain/ccxt/blob/32eb3abd0f226b81955684f17ff9e705e4b86b75/js/pro/bitget.js#L1098
https://github.com/samgermain/ccxt/blob/32eb3abd0f226b81955684f17ff9e705e4b86b75/js/pro/coinbasepro.js#L127
https://github.com/samgermain/ccxt/blob/32eb3abd0f226b81955684f17ff9e705e4b86b75/js/pro/huobi.js#L2100

I know that it doesn't work right, I don't know how to get it to work

Reply:
Hi @samgermain, so that it only calls the function once I recommend using a future as for example here: (https://github.com/pcriadoperez/ccxt/blob/master/js/pro/btcex.js#L778)

On the first call to authenticate, the future won't exist in the client subscriptions, so it will go through the logic of calling watch and authenticating the client.

You will also have to add a function like `handleAuthenticationMessage` that will listen to the response of the auth message and resolve that future. From the docs it looks like this:
```javascript
{
  "data":  {
    "success": true,
    "ts": 1645597033915
  },
  "channel": "auth"
}
```
Once the future is resolved, it is stored in the client.subscriptions so we don't re-authenticate the client every time a private watch function is called.

Let me know if you have any questions, happy to jump on a call and pair program on this one if you need.

Reply:
Why would we use a future instead of authenticating and storing the response object, but then ignoring the authentication step if we already have the response object, kinda like a singleton

Reply:
> it is stored in the client.subscriptions so we don't re-authenticate the client every time a private watch function is called.

How do we access the client from inside the authenticate message? It looks like the client is accessible from the `handle` methods, but not the `watch` methods, which are the ones that would need to call authenticate

Reply:
> Why would we use a future instead of authenticating and storing the response object, but then ignoring the authentication step if we already have the response object, kinda like a singleton

@samgermain first because you don't know when the authentication message will arrive the first time you authenticate and you can't block while you wait. Second if the exchange crashes you don;t want to have pending authentication info in `options`

Reply:
> > it is stored in the client.subscriptions so we don't re-authenticate the client every time a private watch function is called.
> 
> How do we access the client from inside the authenticate message? It looks like the client is accessible from the `handle` methods, but not the `watch` methods, which are the ones that would need to call authenticate

check for instance bybit's authenticate, you do this `        const client = this.client (url)` that will create the client the first time you call it and return it afterwards like a singleton